src/Lerp.cpp

```diff
@@ -134,17 +134,9 @@ Expr lower_lerp(Expr zero_val, Expr one_val, const Expr &weight) {
             case 8:
             case 16:
             case 32: {
-                Expr zero_expand = Cast::make(UInt(2 * bits, computation_type.lanes()),
-                                              zero_val);
-                Expr one_expand = Cast::make(UInt(2 * bits, one_val.type().lanes()),
-                                             one_val);
-
-                Expr rounding = Cast::make(UInt(2 * bits), 1) << Cast::make(UInt(2 * bits), (bits - 1));
-                Expr divisor = Cast::make(UInt(2 * bits), 1) << Cast::make(UInt(2 * bits), bits);
-
-                Expr prod_sum = zero_expand * inverse_typed_weight +
-                                one_expand * typed_weight + rounding;
-                Expr divided = ((prod_sum / divisor) + prod_sum) / divisor;
+                Expr shift = Cast::make(UInt(2 * bits), bits);
+                Expr prod_sum = widening_mul(zero_val, inverse_typed_weight) + widening_mul(one_val, typed_weight);
+                Expr divided = rounding_shift_right(rounding_shift_right(prod_sum, shift) + prod_sum, shift);
```

Comment:
This looks odd. I agree it should be the same as the old code, but I think the inner rounding_shift_right expression will be negligible. @zvookin @abadams any idea what this was about? Is it maybe attempting to ensure we actually hit one_val when typed_weight is the max?

Reply:
It's a trick to do x/255 as (x/256 + x)/256

Reply:
OK yeah, division by 2^n-1...

Reply:
Note that this is a special case of one of our unsigned division lowering methods. I was about to suggest just dividing and relying on that, but that only covers up to 255

Reply:
Actually I could easily extend that table to include 65535 and 2^32 - 1 ...

Reply:
Ah, that's a cool trick, I've been wondering about that too.

Reply:
Hrm, no. This method has the benefit that it doesn't require additional widening.

Reply:
It's not equal! Testing it, I get the following identity instead:

`x/255 == (x/256 + x + 1) / 256`

Reply:
On x86, (x + 127)/255 (which is what I think it should be including rounding), is:

```
	vpaddw	(%rax,%rsi,2), %ymm0, %ymm2
	vpmulhuw	%ymm1, %ymm2, %ymm2
	vpsrlw	$7, %ymm2, %ymm2

```

and ((x + 127)/256 + x + 128) / 256, which is equal, gives:

```
	vpaddw	%ymm0, %ymm2, %ymm3
	vpsrlw	$8, %ymm3, %ymm3
	vpaddw	%ymm3, %ymm2, %ymm2
	vpaddw	%ymm1, %ymm2, %ymm2
	vpsrlw	$8, %ymm2, %ymm2

```

Reply:
So just change that code to something like:
```
Expr divisor = cast(UInt(2 * bits), (1ULL << bits) - 1);
Expr divided = (prod_sum + divisor/2) / divisor;
```

Reply:
Oh nuts, on arm it's reversed:
```
	add	v3.8h, v3.8h, v0.8h
	umull	v4.4s, v3.4h, v1.4h
	umull2	v3.4s, v3.8h, v2.8h
	ushr	v4.4s, v4.4s, #23
	ushr	v3.4s, v3.4s, #23
	xtn	v4.4h, v4.4s
	xtn2	v4.8h, v3.4s
	xtn	v3.8b, v4.8h

```

vs

```
	add	v3.8h, v2.8h, v0.8h
	usra	v2.8h, v3.8h, #8
	addhn	v2.8b, v2.8h, v1.8h
```

Reply:
btw the nasty thing about this trick is that it fails for the largest uint16_t, so you have to have some way to know that's not possible (as we do here).

Reply:
On the other hand, ```x/255 == ((x + 1)/256 + x + 1)/256``` seems to hold for all uint16 x.

So maybe the resolution here is teaching the arm backend this trick for division by 255

Reply:
Do any of us have a copy of the book "Hacker's Delight"? IIRC it has quite a long section on this sort of technique, but I last read it at a different employer ~20 years ago...

Reply:
Here's a useful identity for uint16s:

``` x / 255 === let y = saturating_sub(x, 127) in rounding_shift_right(rounding_shift_right(y, 8) + y, 8)```

On ARM we currently lower x/255 as:

```
	umull	v3.4s, v2.4h, v0.4h
	umull2	v2.4s, v2.8h, v1.8h
	ushr	v3.4s, v3.4s, #23
	ushr	v2.4s, v2.4s, #23
	xtn	v3.4h, v3.4s
	xtn2	v3.8h, v2.4s
	xtn	v2.8b, v3.8h
```

but using this identity we get:

```
	uqsub	v1.8h, v1.8h, v0.8h
	urshr	v2.8h, v1.8h, #8
	raddhn	v1.8b, v2.8h, v1.8h
```

Nice.

Reply:
> Do any of us have a copy of the book "Hacker's Delight"? IIRC it has quite a long section on this sort of technique, but I last read it at a different employer ~20 years ago...

@steven-johnson - very late, I know, but I keep a copy on my desk. If it ever comes up again that we want to consult this book, let me know.